llvm/lib/MC/MCCASObjectV1.cpp

```diff
@@ -1398,6 +1836,77 @@ DebugAbbrevOffsetsRefAdaptor::encodeOffsets(ArrayRef<size_t> Offsets) {
   return EncodedOffsets;
 }
 
+static void
+partitionCUDie(InMemoryCASDWARFObject::PartitionedDebugInfoSection &SplitData,
+               std::unordered_set<llvm::dwarf::Form> &FormsToFixup,
+               DWARFDie &CUDie, ArrayRef<char> DebugInfoData,
+               uint64_t &CUOffset, bool IsLittleEndian, uint8_t AddressSize) {
+
+  // Copy Abbrev Tag
+  DWARFDataExtractor DWARFExtractor(toStringRef(DebugInfoData), IsLittleEndian,
+                                    AddressSize);
+  auto PrevOffset = CUOffset;
+  DWARFExtractor.getULEB128(&CUOffset);
+  auto Size = CUOffset - PrevOffset;
+  for (unsigned I = 0; I < Size; I++)
+    SplitData.DebugInfoRefData.push_back(DebugInfoData[PrevOffset + I]);
+  for (const DWARFAttribute &AttrValue : CUDie.attributes()) {
+    if (FormsToFixup.find(AttrValue.Value.getForm()) != FormsToFixup.end()) {
+      for (unsigned I = 0; I < AttrValue.ByteSize; I++)
+        SplitData.FixupData.push_back(DebugInfoData[AttrValue.Offset + I]);
+    } else {
+      for (unsigned I = 0; I < AttrValue.ByteSize; I++)
+        SplitData.DebugInfoRefData.push_back(
+            DebugInfoData[AttrValue.Offset + I]);
+    }
+    CUOffset += AttrValue.ByteSize;
+  }
+
+  DWARFDie Child = CUDie.getFirstChild();
+  while (Child && Child.getAbbreviationDeclarationPtr()) {
+    partitionCUDie(SplitData, FormsToFixup, Child, DebugInfoData, CUOffset,
+                   IsLittleEndian, AddressSize);
+    Child = Child.getSibling();
+  }
+  if (Child && !Child.getAbbreviationDeclarationPtr()) {
+    SplitData.DebugInfoRefData.push_back(DebugInfoData[CUOffset]);
+    CUOffset++;
+  }
+}
+
+InMemoryCASDWARFObject::PartitionedDebugInfoSection
+InMemoryCASDWARFObject::splitUpCUData(
+    ArrayRef<char> DebugInfoData, uint64_t AbbrevOffset, uint64_t CUOffset,
+    std::unordered_set<llvm::dwarf::Form> &FormsToFixup, DWARFContext *Ctx) {
+
+  StringRef AbbrevSectionContribution =
+      getAbbrevSection().drop_front(AbbrevOffset);
+  DWARFDebugAbbrev Abbrev;
+  Abbrev.extract(DataExtractor(AbbrevSectionContribution, isLittleEndian(), 8));
+  uint64_t OffsetPtr = 0;
+  DWARFUnitHeader Header;
+  DWARFSection Section = {toStringRef(DebugInfoData), 0 /*Address*/};
+  Header.extract(*Ctx, DWARFDataExtractor(*this, Section, isLittleEndian(), 8),
+                 &OffsetPtr, DWARFSectionKind::DW_SECT_INFO);
+
+  DWARFUnitVector UV;
+  DWARFCompileUnit DCU(*Ctx, Section, Header, &Abbrev, &getRangesSection(),
+                       &getLocSection(), getStrSection(),
+                       getStrOffsetsSection(), &getAddrSection(),
+                       getLocSection(), isLittleEndian(), false, UV);
+
+  InMemoryCASDWARFObject::PartitionedDebugInfoSection SplitData;
+  if (DWARFDie CUDie = DCU.getUnitDIE(false)) {
+    // Copy Compile Unit Header.
+    append_range(SplitData.DebugInfoRefData,
+                 DebugInfoData.take_front(DWARF32_HEADER_SIZE));
+    CUOffset += DWARF32_HEADER_SIZE;
```

Comment:
Why do we know the size of the header here and not in `DebugInfoSectionRef::materialize`?